File: app/src/main/java/com/krystianwsul/checkme/gui/TimePickerDialogFragment.java
package com.krystianwsul.checkme.gui;

import android.app.Dialog;
import android.app.TimePickerDialog;
import android.os.Bundle;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.text.format.DateFormat;
import android.widget.TimePicker;

import com.krystianwsul.checkme.utils.time.HourMinute;

import junit.framework.Assert;

public class TimePickerDialogFragment extends AbstractDialogFragment {
    private static final String HOUR_MINUTE_KEY = "hourMinute";

    @Nullable
    private Listener mListener;

    private final TimePickerDialog.OnTimeSetListener mOnTimeSetListener = new TimePickerDialog.OnTimeSetListener() {
        @Override
        public void onTimeSet(TimePicker view, int hourOfDay, int minute) {
            Assert.assertTrue(mListener != null);

            mListener.onHourMinutePicked(new HourMinute(hourOfDay, minute));
        }
    };

    @NonNull
    public static TimePickerDialogFragment newInstance(@NonNull HourMinute hourMinute) {
        TimePickerDialogFragment timePickerDialogFragment = new TimePickerDialogFragment();

        Bundle args = new Bundle();
        args.putParcelable(HOUR_MINUTE_KEY, hourMinute);
        timePickerDialogFragment.setArguments(args);

        return timePickerDialogFragment;
    }

    @NonNull
    @Override
    public Dialog onCreateDialog(Bundle savedInstanceState) {
        Bundle args = getArguments();
        Assert.assertTrue(args != null);
        Assert.assertTrue(args.containsKey(HOUR_MINUTE_KEY));

        HourMinute hourMinute = args.getParcelable(HOUR_MINUTE_KEY);
        Assert.assertTrue(hourMinute != null);

        return new TimePickerDialog(getActivity(), mOnTimeSetListener, hourMinute.getHour(), hourMinute.getMinute(), DateFormat.is24HourFormat(getActivity()));
    }

    public void setListener(@NonNull Listener listener) {
        Assert.assertTrue(mListener == null);

        mListener = listener;
    }

    public interface Listener {
        void onHourMinutePicked(@NonNull HourMinute hourMinute);
    }
}

File: app/src/main/java/com/krystianwsul/checkme/gui/TimePickerDialogFragment.kt
package com.krystianwsul.checkme.gui

import android.app.TimePickerDialog
import android.os.Bundle
import android.text.format.DateFormat
import com.krystianwsul.checkme.utils.time.HourMinute


class TimePickerDialogFragment : AbstractDialogFragment() {

    companion object {

        private const val HOUR_MINUTE_KEY = "hourMinute"

        fun newInstance(hourMinute: HourMinute) = TimePickerDialogFragment().apply {
            arguments = Bundle().apply {
                putParcelable(HOUR_MINUTE_KEY, hourMinute)
            }
        }
    }

    lateinit var listener: (HourMinute) -> Unit

    override fun onCreateDialog(savedInstanceState: Bundle?) = arguments!!.run {
        check(containsKey(HOUR_MINUTE_KEY))

        val hourMinute = getParcelable<HourMinute>(HOUR_MINUTE_KEY)!!

        TimePickerDialog(
                activity,
                TimePickerDialog.OnTimeSetListener { _, hourOfDay, minute -> listener(HourMinute(hourOfDay, minute)) },
                hourMinute.hour,
                hourMinute.minute,
                DateFormat.is24HourFormat(activity))
    }
}
